crates/python/qcs_sdk/qpu/__init__.pyi

```diff
@@ -69,6 +75,27 @@ class QPUResultData:
         Get the mappings of a readout values identifier (ie. "q0") to a set of ``ReadoutValues``
         """
         ...
+    def to_raw_readout_data(
+        self,
+    ) -> RawQPUReadoutData:
+        """
+        Get a copy of this result data flattened into a ``RawQPUReadoutData``
+        """
+        ...
+
+class RawQPUReadoutData:
+    @property
+    def mapping(self) -> Dict[str, str]:
+        """
+        Get the mappings of a memory region (ie. "ro[0]") to it's key name in readout_values
+        """
+        ...
+    @property
+    def readout_values(self) -> Dict[str, Union[List[int], List[float], List[complex]]]:
+        """
+        Get the mappings of a readout values identifier (ie. "q0") to a list of those readout values
```

Comment:
I'm guessing this is something I could learn by reading more about quil, but what are "readout values identifiers"? Are they something the quil program defines, or something internal?

Also, what are the "key names"? Is that internal, or something quil defines?

Are these two maps related/overlapping? I noticed in the tests you have `"a"` as a key for both maps, but based on the comments, it looks like the keys have different meanings.

Reply:
The Quil spec wouldn't answer this question exactly, but as background it is important to understand [Quil-T calibrations](https://quil-lang.github.io/#12-6Defining-Calibrations). They essentially let you map a gate, parameters, and qubits to a sequence of analog control instructions. These instructions offer lower level control than a typical circuit-based program. The default calibrations for a QPU can be retrieved by running `qc.compiler.get_calibration_program()` in pyQuil.

Among other things, these are used to calibrate `MEASURE` instructions on a QPU. So when a user writes something like `MEASURE 0 ro[0]`, the QPU will run a calibration that can actually perform and process the measure. That calibration will write the value to a buffer named something like `q0`. The service maintains a mapping of what readout buffers map to which memory regions in the program, and that mapping gets returned. The key names themselves are defined in the calibrations, so they are defined in Quil, but we control what they are.

The two maps are related. In the above example, a user might want to see the readout values written to `ro[0]` in their program. They would get `mappings["ro[0]"]` as the key in `readout_values` to get the actual values.

Reply:
> The service maintains a mapping of what readout buffers map to which memory regions in the program, and that mapping gets returned

all true but I'd add "...from compilation" to clarify. These mappings are set with the PRAGMA LOAD-MEMORY instruction.

A little more context on this - the data that most users care about (classified qubit values) is actually generated in a post-processing step after execution, and the stream processor is constructed like a tree of processor nodes. Each node has a key, and the "sink" nodes in the graph have the IDs seen as the keys of this readout map as returned from execution. In fact the entire graph is constructed by the user within the Quil program using PRAGMA FILTER-NODE instructions (which themselves come as default in the DEFCAL blocks).

These are great questions from @BatmanAoD and some that many users will ask. This information should get captured in our documentation, likely on gitbook since it's specific to our backend and not to any one client.